// dojo-actions/createAction.js
(function (factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(["require", "exports", 'dojo-compose/compose', 'dojo-compose/mixins/createStateful', 'dojo-core/async/Task', 'dojo-core/WeakMap', './actions'], factory);
    }
})(function (require, exports) {
    "use strict";
    var compose_1 = require('dojo-compose/compose');
    var createStateful_1 = require('dojo-compose/mixins/createStateful');
    var Task_1 = require('dojo-core/async/Task');
    var WeakMap_1 = require('dojo-core/WeakMap');
    var actions_1 = require('./actions');
    /**
     * A type guard that validates the object passed is an Action
     */
    function isAction(value) {
        return typeof value === 'object' && typeof value.do === 'function';
    }
    exports.isAction = isAction;
    /**
     * A weak map of `do` methods
     */
    var doFunctions = new WeakMap_1.default();
    /**
     * A factory which creates instaces of Action
     */
    var createAction = compose_1.default({
        get type() {
            return actions_1.getType(this);
        },
        set type(value) {
            if (!value) {
                throw new TypeError("Cannot set action type to \"" + value + "\"");
            }
            var action = this;
            var currentType = action.type;
            if (currentType && currentType !== value) {
                throw new TypeError("Action type already set as \"" + action.type + "\", cannot change");
            }
            else if (currentType === value) {
                return; /* already added */
            }
            action.own(actions_1.add(action, value));
        },
        do: function (options) {
            var action = this;
            var doFn = doFunctions.get(action);
            if (doFn && action.state.enabled) {
                var result = doFn.call(action, options);
                return Task_1.isTask(result) ? result : Task_1.default.resolve(result);
            }
            return Task_1.default.resolve();
        },
        enable: function () {
            var action = this;
            if (!action.state.enabled) {
                action.setState({ enabled: true });
            }
        },
        disable: function () {
            var action = this;
            if (action.state.enabled) {
                action.setState({ enabled: false });
            }
        }
    })
        .mixin({
        mixin: createStateful_1.default,
        initialize: function (instance, options) {
            if (!options || !options.do) {
                throw new TypeError("'options.do' required during creation.");
            }
            doFunctions.set(instance, options.do);
            instance.setState({ enabled: 'enabled' in options ? options.enabled : true });
            if (options.type) {
                instance.own(actions_1.add(instance, options.type));
            }
            instance.own({
                destroy: function () {
                    doFunctions.delete(instance);
                }
            });
        }
    });
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.default = createAction;
});
